tests/components/tasmota/test_discovery.py

```diff
@@ -0,0 +1,372 @@
+"""The tests for the MQTT discovery."""
+import copy
+import json
+
+from hatasmota.const import CONF_ONLINE
+import pytest
+
+from homeassistant.components.tasmota.const import DEFAULT_PREFIX
+from homeassistant.components.tasmota.discovery import ALREADY_DISCOVERED
+
+from tests.async_mock import patch
+from tests.common import async_fire_mqtt_message
+from tests.components.tasmota.conftest import setup_tasmota_helper
+
+DEFAULT_CONFIG = {
+    "dn": "My Device",
+    "fn": ["Beer", "Milk", "Three", "Four", "Five"],
+    "hn": "tasmota_49A3BC",
+    "mac": "00000049A3BC",
+    "md": "Sonoff 123",
+    "ofl": "offline",
+    CONF_ONLINE: "online",
+    "state": ["OFF", "ON", "TOGGLE", "HOLD"],
+    "sw": "2.3.3.4",
+    "t": "tasmota_49A3BC",
+    "t_f": "%topic%/%prefix%/",
+    "t_p": ["cmnd", "stat", "tele"],
+    "li": [0, 0, 0, 0, 0, 0, 0, 0],
+    "rl": [0, 0, 0, 0, 0, 0, 0, 0],
+    "se": [],
+    "ver": 1,
+}
+
+
+async def test_subscribing_config_topic(hass, mqtt_mock, setup_tasmota):
+    """Test setting up discovery."""
+    discovery_topic = DEFAULT_PREFIX
+
+    assert mqtt_mock.async_subscribe.called
+    call_args = mqtt_mock.async_subscribe.mock_calls[0][1]
+    assert call_args[0] == discovery_topic + "/#"
+    assert call_args[2] == 0
+
+
+async def test_valid_discovery_message(hass, mqtt_mock, caplog):
+    """Test discovery callback called."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+
+    with patch(
+        "homeassistant.components.tasmota.discovery.tasmota_has_entities_with_platform"
+    ) as mock_tasmota_has_entities:
+        await setup_tasmota_helper(hass)
+
+        async_fire_mqtt_message(
+            hass, f"{DEFAULT_PREFIX}/00000049A3BC/config", json.dumps(config)
+        )
+        await hass.async_block_till_done()
+        assert mock_tasmota_has_entities.called
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)